large_image/tilesource/resample.py

```diff
@@ -0,0 +1,124 @@
+from enum import Enum
+from typing import Dict
+
+import numpy as np
+from PIL import Image
+
+
+class ResampleMethod(Enum):
+    PIL_NEAREST = Image.Resampling.NEAREST  # 0
+    PIL_LANCZOS = Image.Resampling.LANCZOS  # 1
+    PIL_BILINEAR = Image.Resampling.BILINEAR  # 2
+    PIL_BICUBIC = Image.Resampling.BICUBIC  # 3
+    PIL_BOX = Image.Resampling.BOX  # 4
+    PIL_HAMMING = Image.Resampling.HAMMING  # 5
+    PIL_MAX_ENUM = 5
+    NP_MEAN = 6
+    NP_MEDIAN = 7
+    NP_MODE = 8
+    NP_MAX = 9
+    NP_MIN = 10
+    NP_NEAREST = 11
+    NP_MAX_CROSSBAND = 12
+    NP_MIN_CROSSBAND = 13
+
+
+def pilResize(
+    tile: np.ndarray,
+    new_shape: Dict,
+    resample_method: ResampleMethod,
+) -> np.ndarray:
+    # Only NEAREST works for 16 bit images
+    img = Image.fromarray(tile)
+    resized_img = img.resize(
+        (new_shape['width'], new_shape['height']),
+        resample=resample_method.value,
+    )
+    result = np.array(resized_img).astype(tile.dtype)
+    return result
+
+
+def numpyResize(
+    tile: np.ndarray,
+    new_shape: Dict,
+    resample_method: ResampleMethod,
+) -> np.ndarray:
+    if resample_method == ResampleMethod.NP_NEAREST:
+        return tile[::2, ::2]
+    else:
+        pixel_selection = None
+        subarrays = np.asarray(
+            [
+                tile[0::2, 0::2],
+                tile[1::2, 0::2],
+                tile[0::2, 1::2],
+                tile[1::2, 1::2],
+            ],
+        )
+
+        if resample_method == ResampleMethod.NP_MEAN:
+            return np.mean(subarrays, axis=0).astype(tile.dtype)
+        elif resample_method == ResampleMethod.NP_MEDIAN:
+            return np.median(subarrays, axis=0).astype(tile.dtype)
+        elif resample_method == ResampleMethod.NP_MODE:
+            # if a pixel occurs twice in a set of four, it is a mode
+            # if no mode, default to pixel 0. check for minimal matches 1=2, 1=3, 2=3
+            pixel_selection = np.where(
+                (
+                    (subarrays[1] == subarrays[2]).all(axis=2) |
+                    (subarrays[1] == subarrays[3]).all(axis=2)
+                ),
+                1, np.where(
+                    (subarrays[2] == subarrays[3]).all(axis=2),
+                    2, 0,
+                ),
+            )
+        elif resample_method == ResampleMethod.NP_MAX:
```

Comment:
I think I was expecting NP_MAX to to be the max of each band independently and NP_MAX_CROSSBAND to be one of the existing pixels.  By my expectation, these are reversed.

Reply:
Oh ok, maybe we can name these better, because I thought "crossband" would mean that the result is a combination of the maxs from each band (and may not be an existing color). 
It does make sense to call the max of each band independently NP_MAX, since it is the simpler version. Maybe we can call the color-preserving version NP_MAX_COLOR or something similar?

Reply:
> Oh ok, maybe we can name these better, because I thought "crossband" would mean that the result is a combination of the maxs from each band (and may not be an existing color). It does make sense to call the max of each band independently NP_MAX, since it is the simpler version. Maybe we can call the color-preserving version NP_MAX_COLOR or something similar?

Other libraries have MAX / MIN and mean the max / min of each band, so we want the simple name to be that.  Do you think NP_MAX_SAMPLE is clearer than NP_MAX_COLOR (pick whichever you like better).

Reply:
We've discovered that "sample" can mean a lot of things, and I feel that "color" implies the combination of bands while "sample" could imply a single band. I think the version that uses the sums to compare pixels is intended to preserve color. I'm leaning towards using `NP_MAX_COLOR` unless you have a preference.

Reply:
Renamed in [ff67307](https://github.com/girder/large_image/pull/1490/commits/ff67307955fbf421b3ab44357dfb5dc1721e25f8)